internal/auth/providers/oidc.go

```diff
@@ -0,0 +1,138 @@
+package providers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"golang.org/x/oauth2"
+
+	log "github.com/buzzfeed/sso/internal/pkg/logging"
+	"github.com/buzzfeed/sso/internal/pkg/sessions"
+	oidc "github.com/coreos/go-oidc"
+)
+
+// OIDCProvider is a generic OpenID Connect provider
+type OIDCProvider struct {
+	*ProviderData
+
+	Verifier *oidc.IDTokenVerifier
+}
+
+// NewOIDCProvider creates a new generic OpenID Connect provider
+func NewOIDCProvider(p *ProviderData) *OIDCProvider {
+	p.ProviderName = "OpenID Connect"
+	return &OIDCProvider{ProviderData: p}
+}
+
+// Redeem fulfills the Provider interface.
+// The authenticator uses this method to redeem the code provided to /callback after the user logs into their OpenID Connect account.
+func (p *OIDCProvider) Redeem(redirectURL, code string) (s *sessions.SessionState, err error) {
```

Comment:
It's been our experience that named return parameters more frequently obfuscate control flow than help to clarify it, for later readers of the code. I don't necessarily think that's the case here, but we consequently make it a general rule to avoid them, except in circumstances where they're necessary (e.g. modifying returned arguments within a deferred function call).